src/hotspot/share/prims/whitebox.cpp

```diff
@@ -1938,49 +1940,73 @@ WB_ENTRY(jobjectArray, WB_getObjectsViaKlassOopMaps(JNIEnv* env, jobject wb, job
   return (jobjectArray)JNIHandles::make_local(THREAD, result_array());
 WB_END
 
-class CollectOops : public BasicOopIterateClosure {
- public:
-  GrowableArray<Handle>* array;
+// Collect Object oops but not value objects...loaded from heap
+class CollectObjectOops : public BasicOopIterateClosure {
+  public:
+  GrowableArray<Handle>* _array;
+
+  CollectObjectOops() {
+      _array = new GrowableArray<Handle>(128);
+  }
+
+  void add_oop(Handle oh) {
 
-  jobjectArray create_jni_result(JNIEnv* env, TRAPS) {
-    objArrayHandle result_array =
-        oopFactory::new_objArray_handle(vmClasses::Object_klass(), array->length(), CHECK_NULL);
-    for (int i = 0 ; i < array->length(); i++) {
-      result_array->obj_at_put(i, array->at(i)());
-    }
-    return (jobjectArray)JNIHandles::make_local(THREAD, result_array());
   }
 
   void add_oop(oop o) {
     Handle oh = Handle(Thread::current(), o);
-    // Value might be oop, but JLS can't see as Object, just iterate through it...
     if (oh != nullptr && oh->is_inline_type()) {
       oh->oop_iterate(this);
     } else {
-      array->append(oh);
+      _array->append(oh);
     }
   }
 
   void do_oop(oop* o) { add_oop(HeapAccess<>::oop_load(o)); }
   void do_oop(narrowOop* v) { add_oop(HeapAccess<>::oop_load(v)); }
+
+  jobjectArray create_jni_result(JNIEnv* env, TRAPS) {
+    objArrayHandle result_array =
+        oopFactory::new_objArray_handle(vmClasses::Object_klass(), _array->length(), CHECK_NULL);
+    for (int i = 0 ; i < _array->length(); i++) {
+      result_array->obj_at_put(i, _array->at(i)());
+    }
+    return (jobjectArray)JNIHandles::make_local(THREAD, result_array());
+  }
+};
+
+// Collect Object oops but not value objects...loaded from frames
+class CollectFrameObjectOops : public BasicOopIterateClosure {
+ public:
+  CollectObjectOops _collect;
+
+  void add_oop(oop o) {
+    _collect.add_oop(o);
+  }
+
+  void do_oop(oop* o) { add_oop(*o); }
+  void do_oop(narrowOop* v) { add_oop(CompressedOops::decode(*v)); }
```

Comment:
I think we rarely do this directly but let RawAccess handle the load and decode. 
```suggestion
  void do_oop(oop* o) { add_oop(RawAccess<>::oop_load(o)); }
  void do_oop(narrowOop* v) { add_oop(RawAccess<>::oop_load(v)); }
```

and the most common way to write oop closures which are used in both oop and narrowOop contexts:
```c++
  template <class T>
  void do_oop_work(T* p) {
    oop o =  RawAccess<>::oop_load(p); 
    // Do the work
  }
  void do_oop(oop* o) { do_oop_work(o); }
  void do_oop(narrowOop* v) { do_oop_work(v); }
```

Reply:
You are correct, I want to thank @fisk for leading me astray on this one :-) Fixed it up.